File: SampleConsensusPrerejective.lua
local ffi = require 'ffi'
local torch = require 'torch'
local utils = require 'pcl.utils'
local pcl = require 'pcl.PointTypes'

local SampleConsensusPrerejective = torch.class('pcl.SampleConsensusPrerejective', pcl)

local func_by_type = {}

local function init()
  local SampleConsensusPrerejective_method_names = {
    'new',
    'delete',
    'setInputSource',
    'setSourceFeatures',
    'setInputTarget',
    'setTargetFeatures',
    'setMaximumIterations',
    'getMaximumIterations',
    'setNumberOfSamples',
    'getNumberOfSamples',
    'getMaxCorrespondenceDistance',
    'setMaxCorrespondenceDistance',
    'getCorrespondenceRandomness',
    'setCorrespondenceRandomness',
    'setSimilarityThreshold',
    'getSimilarityThreshold',
    'setInlierFraction',
    'getInlierFraction',
    'getInliers',
    'getFinalTransformation',
    'getFitnessScore',
    'align'
  }
  for k,v in pairs(utils.type_key_map) do
    func_by_type[k] = utils.create_typed_methods("pcl_SampleConsensusPrerejective_TYPE_KEY_", SampleConsensusPrerejective_method_names, v)
  end
end

init()

function SampleConsensusPrerejective:__init(pointType)
  self.pointType = pcl.pointType(pointType or pcl.PointXYZ)
  self.f = func_by_type[self.pointType]
  self.o = self.f.new()
end

function SampleConsensusPrerejective:cdata()
  return self.o
end

function SampleConsensusPrerejective:setInputSource(source_cloud)
  self.f.setInputSource(self.o, source_cloud:cdata())
end

function SampleConsensusPrerejective:setSourceFeatures(source_features)
  self.f.setSourceFeatures(self.o, source_features:cdata())
end

function SampleConsensusPrerejective:setInputTarget(target_cloud)
  self.f.setInputTarget(self.o, target_cloud:cdata())
end

function SampleConsensusPrerejective:setTargetFeatures(target_features)
  self.f.setTargetFeatures(self.o, target_features:cdata())
end

function SampleConsensusPrerejective:setMaximumIterations(max_iterations)
  self.f.setMaximumIterations(self.o, max_iterations)
end

function SampleConsensusPrerejective:getMaximumIterations()
  return self.f.getMaximumIterations(self.o)
end

function SampleConsensusPrerejective:setNumberOfSamples(nr_samples)
  self.f.setNumberOfSamples(self.o, nr_samples)
end

function SampleConsensusPrerejective:getNumberOfSamples()
  return self.f.getNumberOfSamples(self.o)
end

function SampleConsensusPrerejective:getMaxCorrespondenceDistance()
  return self.f.getMaxCorrespondenceDistance(self.o)
end

function SampleConsensusPrerejective:setMaxCorrespondenceDistance(distance)
  self.f.setMaxCorrespondenceDistance(self.o, distance)
end

function SampleConsensusPrerejective:setCorrespondenceRandomness(k)
  self.f.setCorrespondenceRandomness(self.o, k)
end

function SampleConsensusPrerejective:getCorrespondenceRandomness()
  return  self.f.getCorrespondenceRandomness(self.o)
end

function SampleConsensusPrerejective:setSimilarityThreshold(similarity_threshold)
  self.f.setSimilarityThreshold(self.o, similarity_threshold)
end

function SampleConsensusPrerejective:getSimilarityThreshold()
  return  self.f.getSimilarityThreshold(self.o)
end

function SampleConsensusPrerejective:setInlierFraction(inlier_fraction)
  self.f.setInlierFraction(self.o, inlier_fraction)
end

function SampleConsensusPrerejective:getInlierFraction()
  return self.f.getInlierFraction(self.o)
end

function SampleConsensusPrerejective:getInliers(output)
  output = output or pcl.Indices()
  self.f.getInliers(self.o, output:cdata()) 
  return output
end

function SampleConsensusPrerejective:getFinalTransformation()
  local t = torch.FloatTensor()
  self.f.getFinalTransformation(self.o, t:cdata())
  return t
end

function SampleConsensusPrerejective:getFitnessScore(max_range)
  return self.f.getFitnessScore(self.o, max_range or pcl.range.double.max)
end

function SampleConsensusPrerejective:align(output, initial_guess)
  output = output or pcl.PointCloud(self.pointType)
  if initial_guess then
    initial_guess = initial_guess:cdata()
  end
  self.f.align(self.o, output:cdata(), initial_guess)
  return output
end

File: src/generic/sample_consensus_prerejective.cpp
#include <pcl/registration/sample_consensus_prerejective.h>

#define SampleConsensusPrerejective_ptr pcl::SampleConsensusPrerejective<_PointT, _PointT, pcl::FPFHSignature33>::Ptr
#define PointCloud_ptr pcl::PointCloud<_PointT>::Ptr
#define FeatureCloud_ptr pcl::PointCloud<pcl::FPFHSignature33>::Ptr

PCLIMP(SampleConsensusPrerejective_ptr *, SampleConsensusPrerejective, new)()
{
  return new SampleConsensusPrerejective_ptr(new pcl::SampleConsensusPrerejective<_PointT, _PointT, pcl::FPFHSignature33>());
}

PCLIMP(void, SampleConsensusPrerejective, delete)(SampleConsensusPrerejective_ptr *self)
{
  delete self;
}

PCLIMP(void, SampleConsensusPrerejective, setInputSource)(SampleConsensusPrerejective_ptr *self, PointCloud_ptr *source_cloud)
{
  (*self)->setInputSource(*source_cloud);
}

PCLIMP(void, SampleConsensusPrerejective, setSourceFeatures)(SampleConsensusPrerejective_ptr *self, FeatureCloud_ptr* source_features)
{
  (*self)->setSourceFeatures(*source_features);
}

PCLIMP(void, SampleConsensusPrerejective, setInputTarget)(SampleConsensusPrerejective_ptr *self, PointCloud_ptr *target_cloud)
{
  (*self)->setInputTarget(*target_cloud);
}

PCLIMP(void, SampleConsensusPrerejective, setTargetFeatures)(SampleConsensusPrerejective_ptr *self, FeatureCloud_ptr* target_features)
{
  (*self)->setTargetFeatures(*target_features);
}

PCLIMP(void, SampleConsensusPrerejective, setMaximumIterations)(SampleConsensusPrerejective_ptr *self, int max_iterations)
{
  (*self)->setMaximumIterations(max_iterations);
}

PCLIMP(int, SampleConsensusPrerejective, getMaximumIterations)(SampleConsensusPrerejective_ptr *self)
{
  return (*self)->getMaximumIterations();
}

PCLIMP(void, SampleConsensusPrerejective, setNumberOfSamples)(SampleConsensusPrerejective_ptr *self, int nr_samples)
{
  (*self)->setNumberOfSamples(nr_samples);
}

PCLIMP(int, SampleConsensusPrerejective, getNumberOfSamples)(SampleConsensusPrerejective_ptr *self)
{
  return (*self)->getNumberOfSamples();
}

PCLIMP(void, SampleConsensusPrerejective, setMaxCorrespondenceDistance)(SampleConsensusPrerejective_ptr *self, double distance)
{
  (*self)->setMaxCorrespondenceDistance(distance);
}

PCLIMP(double, SampleConsensusPrerejective, getMaxCorrespondenceDistance)(SampleConsensusPrerejective_ptr *self)
{
  return (*self)->getMaxCorrespondenceDistance();
}

PCLIMP(void, SampleConsensusPrerejective, setCorrespondenceRandomness)(SampleConsensusPrerejective_ptr *self, int k)
{
  (*self)->setCorrespondenceRandomness(k);
}

PCLIMP(int, SampleConsensusPrerejective, getCorrespondenceRandomness)(SampleConsensusPrerejective_ptr *self)
{
  return (*self)->getCorrespondenceRandomness();
}

PCLIMP(void, SampleConsensusPrerejective, setSimilarityThreshold)(SampleConsensusPrerejective_ptr *self, float similarity_threshold)
{
  (*self)->setSimilarityThreshold(similarity_threshold);
}

PCLIMP(float, SampleConsensusPrerejective, getSimilarityThreshold)(SampleConsensusPrerejective_ptr *self)
{
  return (*self)->getSimilarityThreshold();
}

PCLIMP(void, SampleConsensusPrerejective, setInlierFraction)(SampleConsensusPrerejective_ptr *self, float inlier_fraction)
{
  (*self)->setInlierFraction(inlier_fraction);
}

PCLIMP(float, SampleConsensusPrerejective, getInlierFraction)(SampleConsensusPrerejective_ptr *self)
{
  return (*self)->getInlierFraction();
}

PCLIMP(void, SampleConsensusPrerejective, getInliers)(SampleConsensusPrerejective_ptr *self, Indices_ptr *indices)
{
  std::vector<int>& out = **indices;
  out = (*self)->getInliers();
}

PCLIMP(void, SampleConsensusPrerejective, getFinalTransformation)(SampleConsensusPrerejective_ptr *self, THFloatTensor* output)
{
  Eigen::Matrix4f transformation = (*self)->getFinalTransformation();
  copyMatrix(transformation, output);
}

PCLIMP(double, SampleConsensusPrerejective, getFitnessScore)(SampleConsensusPrerejective_ptr *self, double max_range)
{
  return (*self)->getFitnessScore(max_range);
}

PCLIMP(void, SampleConsensusPrerejective, align)(SampleConsensusPrerejective_ptr *self, PointCloud_ptr *output, THFloatTensor *guess)
{
  if (!guess)
    (*self)->align(**output);
  else
    (*self)->align(**output, Tensor2Mat<4,4>(guess));
}

#undef SampleConsensusPrerejective_ptr
#undef PointCloud_ptr
